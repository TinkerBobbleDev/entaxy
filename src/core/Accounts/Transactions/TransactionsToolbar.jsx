import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'
import IconButton from '@material-ui/core/IconButton'
import Tooltip from '@material-ui/core/Tooltip'
import AddIcon from '@material-ui/icons/Add'
import Icon from '@mdi/react'
import DeleteIcon from '@material-ui/icons/Delete'
import { mdiImport } from '@mdi/js'
import InputBase from '@material-ui/core/InputBase'
import InputAdornment from '@material-ui/core/InputAdornment'
import SearchIcon from '@material-ui/icons/Search'
import { fade } from '@material-ui/core/styles/colorManipulator'
import confirm from '../../../util/confirm'
import TableToolbar from '../../../common/TableToolbar'
import LinkTo from '../../../common/LinkTo'

const styles = (theme) => ({
  importButton: {
    fill: theme.palette.text.secondary
  },
  buttons: {
    display: 'flex'
  },
  search: {
    position: 'relative',
    borderRadius: theme.shape.borderRadius * 2,
    backgroundColor: fade(theme.palette.grey[400], 0.15),
    marginRight: theme.spacing(2),
    marginLeft: 0,
    width: '100%',
    '&:hover': {
      backgroundColor: fade(theme.palette.grey[400], 0.25)
    }
  },
  inputRoot: {
    color: 'inherit',
    width: '100%',
    padding: theme.spacing(2),
    paddingTop: theme.spacing(1),
    paddingBottom: theme.spacing(1)

  },
  inputInput: {
    transition: theme.transitions.create('width'),
    width: '100%',
    [theme.breakpoints.up('md')]: {
      minWidth: 240
    }
  }
})

export class TransactionsToolbarComponent extends React.Component {
  onDelete = () => {
    confirm('Delete selected transactions?', 'Are you sure?').then(() => {
      this.props.handleDelete(this.props.selectedTransactions)
      this.props.resetSelection()
    })
  }

  onChangeSearch = ({ target: { value } }) => {
    const { filterProps } = this.props
    if (value !== '') {
      filterProps.setFilter({
        attr: 'description',
        value: value.toLowerCase()
      })
    } else {
      filterProps.unsetFilter({ attr: 'description' })
    }
  }

  render() {
    const {
      classes,
      account,
      handleNew,
      selectedTransactions,
      filterProps
    } = this.props

    return (
      <TableToolbar
        title={`${account.institution} - ${account.name}`}
        subTitle={account.currency}
        selectedItems={selectedTransactions}
      >
        {selectedTransactions.length > 0 ? (
          <Tooltip title="Delete">
            <IconButton aria-label="Delete" onClick={this.onDelete}>
              <DeleteIcon />
            </IconButton>
          </Tooltip>
        ) : (
          <div className={classes.buttons}>
            <div className={classes.search}>
              <InputBase
                type="search"
                placeholder="Search transactions"
                onChange={this.onChangeSearch}
                value={filterProps.filters.description || ''}
                classes={{
                  root: classes.inputRoot,
                  input: classes.inputInput
                }}
                startAdornment={<InputAdornment position="start"><SearchIcon /></InputAdornment>}
              />
            </div>
            <Tooltip title="Import transaction">
              <IconButton
                aria-label="Import transaction"
                component={LinkTo(`/accounts/${account.id}/import`)}
              >
                <Icon
                  path={mdiImport}
                  size={1}
                  className={classes.importButton}
                />
              </IconButton>
            </Tooltip>
            <Tooltip title="New transaction">
              <IconButton aria-label="New transaction" onClick={handleNew}>
                <AddIcon />
              </IconButton>
            </Tooltip>
          </div>
        )}
      </TableToolbar>
    )
  }
}

TransactionsToolbarComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  account: PropTypes.object.isRequired,
  handleNew: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
  selectedTransactions: PropTypes.array.isRequired,
  resetSelection: PropTypes.func.isRequired,
  filterProps: PropTypes.object.isRequired
}

export default withStyles(styles)(withRouter(TransactionsToolbarComponent))
