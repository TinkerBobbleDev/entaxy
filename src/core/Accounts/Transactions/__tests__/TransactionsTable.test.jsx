import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { BrowserRouter } from 'react-router-dom'
import { TransactionsTableComponent } from '../TransactionsTable'
import TransactionsToolbar from '../TransactionsToolbar'

describe('TransactionsTable', () => {
  const account = { id: 1, name: 'Checking', institution: 'TD' }
  const transactions = [
    { id: 1, description: 'Transaction 1', createdAt: Date.now() },
    { id: 2, description: 'Transaction 2', createdAt: Date.now() + 10 },
    { id: 3, description: 'Transaction 3', createdAt: Date.now() + 20 }
  ]
  const mochHandleNew = jest.fn()
  const mochHandleDelete = jest.fn()
  const mochFormatCurrency = jest.fn().mockReturnValue(() => {})
  const mochFormatDecimal = jest.fn().mockReturnValue(() => {})
  const mochFormatDate = jest.fn().mockReturnValue(() => {})
  let wrapper
  let instance

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('matches snapshot with no transactions', () => {
    const component = renderer.create((
      <BrowserRouter>
        <TransactionsTableComponent
          classes={{}}
          className=""
          account={account}
          transactions={[]}
          budget={{}}
          formatCurrency={mochFormatCurrency}
          formatDecimal={mochFormatDecimal}
          formatDate={mochFormatDate}
          Toolbar={TransactionsToolbar}
          toolbarProps={{
            handleNew: mochHandleNew,
            handleDelete: mochHandleDelete
          }}
        />
      </BrowserRouter>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('matches snapshot with some transactions', () => {
    const component = renderer.create((
      <BrowserRouter>
        <TransactionsTableComponent
          classes={{}}
          className=""
          account={account}
          transactions={transactions}
          budget={{}}
          formatCurrency={mochFormatCurrency}
          formatDecimal={mochFormatDecimal}
          formatDate={mochFormatDate}
          Toolbar={TransactionsToolbar}
          toolbarProps={{
            handleNew: mochHandleNew,
            handleDelete: mochHandleDelete
          }}
        />
      </BrowserRouter>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  describe('Component methods', () => {
    beforeEach(() => {
      wrapper = shallow((
        <TransactionsTableComponent
          classes={{}}
          className=""
          account={account}
          transactions={transactions}
          budget={{}}
          formatCurrency={mochFormatCurrency}
          formatDecimal={mochFormatDecimal}
          formatDate={mochFormatDate}
          Toolbar={TransactionsToolbar}
          toolbarProps={{
            handleNew: mochHandleNew,
            handleDelete: mochHandleDelete
          }}
        />
      ))
      instance = wrapper.instance()
    })

    describe('handleSort', () => {
      it('should sort ASC and DESC', () => {
        expect(instance.state.sortBy).toEqual('createdAt')
        expect(instance.state.sortDirection).toEqual('DESC')

        instance.handleSort({ sortBy: 'createdAt', sortDirection: 'ASC' })
        expect(instance.state.sortBy).toEqual('createdAt')
        expect(instance.state.sortDirection).toEqual('ASC')
      })
    })

    describe('filterAndSortTransactions', () => {
      it('should do basic orderBy', () => {
        instance.setState({
          sortBy: 'createdAt',
          sortDirection: 'ASC'
        })
        expect(instance.filterAndSortTransactions()).toEqual(transactions)
        instance.setState({
          sortBy: 'createdAt',
          sortDirection: 'DESC'
        })
        expect(instance.filterAndSortTransactions()).toEqual([...transactions].reverse())
      })

      it('should do basic filter', () => {
        instance.setState({
          filters: {
            createdAt: transactions[0].createdAt
          }
        })
        expect(instance.filterAndSortTransactions()).toEqual([transactions[0]])
        instance.setState({
          filters: {
            createdAt: transactions[1].createdAt,
            description: 'Bogus'
          }
        })
        expect(instance.filterAndSortTransactions()).toEqual([])
        instance.setState({
          filters: {
            createdAt: transactions[1].createdAt,
            description: transactions[1].description.toLowerCase()
          }
        })
        expect(instance.filterAndSortTransactions()).toEqual([transactions[1]])
      })

      it('should filter with a function', () => {
        instance.setState({
          sortBy: 'createdAt',
          sortDirection: 'ASC',
          filters: {
            createdAt: t => t.createdAt > transactions[0].createdAt
          }
        })
        expect(instance.filterAndSortTransactions()).toEqual([transactions[1], transactions[2]])
      })
    })

    describe('rowClassName', () => {
      const classes = {
        headerRow: 'headerRow',
        nativeAmount: 'nativeAmount',
        oddRow: 'oddRow',
        row: 'row',
        rowWithError: 'rowWithError',
        tableWrapper: 'tableWrapper'
      }

      it('should select row classes', () => {
        expect(instance.rowClassName({ index: -1 }, transactions, classes)).toEqual('headerRow oddRow')
        expect(instance.rowClassName({ index: 0 }, transactions, classes)).toEqual('row')
        expect(instance.rowClassName({ index: 1 }, transactions, classes)).toEqual('row oddRow')
        transactions[2].errors = ['Some error']
        expect(instance.rowClassName({ index: 2 }, transactions, classes)).toEqual('rowWithError row')
      })
    })

    describe('renderCellAmount', () => {
      it('should render for positive amounts', () => {
        expect(instance.renderCellAmount({ cellData: { amount: 1, restrictTo: 'positiveAmount' } })).not.toBeNull()
        expect(instance.renderCellAmount({ cellData: { amount: 0, restrictTo: 'positiveAmount' } })).toBeNull()
        expect(instance.renderCellAmount({ cellData: { amount: -1, restrictTo: 'positiveAmount' } })).toBeNull()
      })

      it('should render for negative amounts', () => {
        expect(instance.renderCellAmount({ cellData: { amount: -11, restrictTo: 'negativeAmount' } })).not.toBeNull()
        expect(instance.renderCellAmount({ cellData: { amount: 0, restrictTo: 'negativeAmount' } })).toBeNull()
        expect(instance.renderCellAmount({ cellData: { amount: 1, restrictTo: 'negativeAmount' } })).toBeNull()
      })
    })

    describe('setFilter, unsetFilter and resetFilters', () => {
      it('should set and unset filters on transactions', () => {
        expect(instance.state).toEqual({
          selected: [],
          filters: {},
          sortBy: 'createdAt',
          sortDirection: 'DESC',
          prevPropsAccountId: account.id
        })

        // Bogus filter
        instance.setFilter({ attr: 'filter1', value: 1 })
        expect(instance.state.filters).toEqual({ filter1: 1 })
        // Same filter again
        instance.setFilter({ attr: 'filter1', value: 1 })
        expect(instance.state.filters).toEqual({ filter1: 1 })
        // New filter
        instance.setFilter({ attr: 'id', value: 2 })
        expect(instance.state.filters).toEqual({ filter1: 1, id: 2 })
        // Remove the first filter
        instance.unsetFilter({ attr: 'filter1' })
        expect(instance.state.filters).toEqual({ id: 2 })
        // Reset all filter
        instance.resetFilters()
        expect(instance.state.filters).toEqual({ })
      })
    })

    describe('Transactions selection', () => {
      it('selects a single transaction', () => {
        expect(wrapper.state('selected')).toEqual([])
        // Select one
        instance.handleCheckboxClick(null, 1)
        expect(wrapper.state('selected')).toEqual([1])
        // Select another
        instance.handleCheckboxClick(null, 2)
        expect(wrapper.state('selected')).toEqual([1, 2])
        // Select one more
        instance.handleCheckboxClick(null, 4)
        expect(wrapper.state('selected')).toEqual([1, 2, 4])
        // Select one more
        instance.handleCheckboxClick(null, 3)
        expect(wrapper.state('selected')).toEqual([1, 2, 4, 3])

        // Un-select one from the middle
        instance.handleCheckboxClick(null, 4)
        expect(wrapper.state('selected')).toEqual([1, 2, 3])
        // Un-select last one
        instance.handleCheckboxClick(null, 3)
        expect(wrapper.state('selected')).toEqual([1, 2])
        // Un-select first one
        instance.handleCheckboxClick(null, 1)
        expect(wrapper.state('selected')).toEqual([2])
        // Empty
        instance.handleCheckboxClick(null, 2)
        expect(wrapper.state('selected')).toEqual([])
      })

      it('selects all transactions', () => {
        expect(wrapper.state('selected')).toEqual([])
        instance.handleSelectAllClick({ target: { checked: true } }, transactions)
        expect(wrapper.state('selected')).toEqual(transactions.map(n => n.id))
      })

      it('check selection', () => {
        expect(wrapper.state('selected')).toEqual([])
        instance.handleCheckboxClick(null, 1)
        expect(instance.isSelected(1)).toBe(true)
        expect(instance.isSelected(10)).toBe(false)
      })

      it('de-selects all transactions', () => {
        expect(wrapper.state('selected')).toEqual([])
        instance.handleCheckboxClick(null, 1)
        instance.handleCheckboxClick(null, 2)
        expect(wrapper.state('selected')).toEqual([1, 2])
        instance.handleSelectAllClick({ target: { checked: false } }, transactions)
        expect(wrapper.state('selected')).toEqual([])
      })
    })
  })
})
