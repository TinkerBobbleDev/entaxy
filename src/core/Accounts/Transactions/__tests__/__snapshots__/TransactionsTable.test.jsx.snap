// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsTable matches snapshot with no transactions 1`] = `
<div
  className=" "
>
  <div
    className="MuiToolbar-root-8 MuiToolbar-regular-10 MuiToolbar-gutters-9 TableToolbar-root-3"
  >
    <div
      className="TableToolbar-title-7"
    >
      <div>
        <h6
          className="MuiTypography-root-12 MuiTypography-h6-29"
        >
          TD - Checking
        </h6>
        <span
          className="MuiTypography-root-12 MuiTypography-caption-22"
        />
      </div>
    </div>
    <div
      className="TableToolbar-spacer-5"
    />
    <div
      className="TableToolbar-actions-6"
    >
      <div
        className="withRouter-TransactionsToolbarComponent--buttons-2"
      >
        <a
          aria-current={null}
          aria-describedby={null}
          aria-label="Import transaction"
          className="MuiButtonBase-root-61 MuiIconButton-root-55"
          href="/accounts/1/import"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
          title="Import transaction"
        >
          <span
            className="MuiIconButton-label-60"
          >
            <svg
              className="withRouter-TransactionsToolbarComponent--importButton-1"
              style={
                Object {
                  "width": "1.5rem",
                }
              }
              viewBox="0 0 24 24"
            >
              <path
                d="M14,12L10,8V11H2V13H10V16M20,18V6C20,4.89 19.1,4 18,4H6C4.9,4 4,4.9 4,6V9H6V6H18V18H6V15H4V18C4,19.1 4.9,20 6,20H18C19.1,20 20,19.1 20,18Z"
                style={Object {}}
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-83"
          />
        </a>
        <button
          aria-describedby={null}
          aria-label="New transaction"
          className="MuiButtonBase-root-61 MuiIconButton-root-55"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          title="New transaction"
          type="button"
        >
          <span
            className="MuiIconButton-label-60"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-64"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-83"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  >
    <div
      aria-colcount={5}
      aria-rowcount={0}
      className="ReactVirtualized__Table"
      role="grid"
      style={Object {}}
    >
      <div
        className="ReactVirtualized__Table__headerRow undefined"
        role="row"
        style={
          Object {
            "height": 25,
            "overflow": "hidden",
            "paddingRight": 0,
            "width": 0,
          }
        }
      >
        <div
          className="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 40px",
              "flex": "0 1 40px",
              "msFlex": "0 1 40px",
            }
          }
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
          >
            <span
              className="MuiButtonBase-root-61 MuiIconButton-root-55 MuiPrivateSwitchBase-root-79 MuiCheckbox-root-73 MuiCheckbox-colorSecondary-78"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label-60"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-64"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
                <input
                  checked={false}
                  className="MuiPrivateSwitchBase-input-82"
                  data-indeterminate={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </span>
              <span
                className="MuiTouchRipple-root-83"
              />
            </span>
          </span>
        </div>
        <div
          aria-label="Date"
          aria-sort="descending"
          className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 120px",
              "flex": "0 1 120px",
              "msFlex": "0 1 120px",
            }
          }
          tabIndex={0}
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="Date"
          >
            Date
          </span>
          <svg
            className="ReactVirtualized__Table__sortableHeaderIcon ReactVirtualized__Table__sortableHeaderIcon--DESC"
            height={18}
            viewBox="0 0 24 24"
            width={18}
          >
            <path
              d="M7 10l5 5 5-5z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <div
          className="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "1 1 200px",
              "flex": "1 1 200px",
              "msFlex": "1 1 200px",
            }
          }
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="Description"
          >
            Description
          </span>
        </div>
        <div
          aria-label="In"
          aria-sort="none"
          className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 130px",
              "flex": "0 1 130px",
              "msFlex": "0 1 130px",
            }
          }
          tabIndex={0}
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="In"
          >
            In
          </span>
        </div>
        <div
          aria-label="Out"
          aria-sort="none"
          className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 130px",
              "flex": "0 1 130px",
              "msFlex": "0 1 130px",
            }
          }
          tabIndex={0}
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="Out"
          >
            Out
          </span>
        </div>
      </div>
      <div
        aria-label="grid"
        aria-readonly={true}
        className="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
        onScroll={[Function]}
        role="rowgroup"
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "boxSizing": "border-box",
            "direction": "ltr",
            "height": -25,
            "overflowX": "hidden",
            "overflowY": "auto",
            "position": "relative",
            "width": 0,
            "willChange": "transform",
          }
        }
        tabIndex={0}
      />
    </div>
  </div>
</div>
`;

exports[`TransactionsTable matches snapshot with some transactions 1`] = `
<div
  className=" "
>
  <div
    className="MuiToolbar-root-8 MuiToolbar-regular-10 MuiToolbar-gutters-9 TableToolbar-root-3"
  >
    <div
      className="TableToolbar-title-7"
    >
      <div>
        <h6
          className="MuiTypography-root-12 MuiTypography-h6-29"
        >
          TD - Checking
        </h6>
        <span
          className="MuiTypography-root-12 MuiTypography-caption-22"
        />
      </div>
    </div>
    <div
      className="TableToolbar-spacer-5"
    />
    <div
      className="TableToolbar-actions-6"
    >
      <div
        className="withRouter-TransactionsToolbarComponent--buttons-2"
      >
        <a
          aria-current={null}
          aria-describedby={null}
          aria-label="Import transaction"
          className="MuiButtonBase-root-61 MuiIconButton-root-55"
          href="/accounts/1/import"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
          title="Import transaction"
        >
          <span
            className="MuiIconButton-label-60"
          >
            <svg
              className="withRouter-TransactionsToolbarComponent--importButton-1"
              style={
                Object {
                  "width": "1.5rem",
                }
              }
              viewBox="0 0 24 24"
            >
              <path
                d="M14,12L10,8V11H2V13H10V16M20,18V6C20,4.89 19.1,4 18,4H6C4.9,4 4,4.9 4,6V9H6V6H18V18H6V15H4V18C4,19.1 4.9,20 6,20H18C19.1,20 20,19.1 20,18Z"
                style={Object {}}
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-83"
          />
        </a>
        <button
          aria-describedby={null}
          aria-label="New transaction"
          className="MuiButtonBase-root-61 MuiIconButton-root-55"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          title="New transaction"
          type="button"
        >
          <span
            className="MuiIconButton-label-60"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-64"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-83"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  >
    <div
      aria-colcount={5}
      aria-rowcount={3}
      className="ReactVirtualized__Table"
      role="grid"
      style={Object {}}
    >
      <div
        className="ReactVirtualized__Table__headerRow undefined"
        role="row"
        style={
          Object {
            "height": 25,
            "overflow": "hidden",
            "paddingRight": 0,
            "width": 0,
          }
        }
      >
        <div
          className="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 40px",
              "flex": "0 1 40px",
              "msFlex": "0 1 40px",
            }
          }
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
          >
            <span
              className="MuiButtonBase-root-61 MuiIconButton-root-55 MuiPrivateSwitchBase-root-79 MuiCheckbox-root-73 MuiCheckbox-colorSecondary-78"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label-60"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-64"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
                <input
                  checked={false}
                  className="MuiPrivateSwitchBase-input-82"
                  data-indeterminate={false}
                  onChange={[Function]}
                  type="checkbox"
                />
              </span>
              <span
                className="MuiTouchRipple-root-83"
              />
            </span>
          </span>
        </div>
        <div
          aria-label="Date"
          aria-sort="descending"
          className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 120px",
              "flex": "0 1 120px",
              "msFlex": "0 1 120px",
            }
          }
          tabIndex={0}
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="Date"
          >
            Date
          </span>
          <svg
            className="ReactVirtualized__Table__sortableHeaderIcon ReactVirtualized__Table__sortableHeaderIcon--DESC"
            height={18}
            viewBox="0 0 24 24"
            width={18}
          >
            <path
              d="M7 10l5 5 5-5z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <div
          className="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "1 1 200px",
              "flex": "1 1 200px",
              "msFlex": "1 1 200px",
            }
          }
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="Description"
          >
            Description
          </span>
        </div>
        <div
          aria-label="In"
          aria-sort="none"
          className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 130px",
              "flex": "0 1 130px",
              "msFlex": "0 1 130px",
            }
          }
          tabIndex={0}
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="In"
          >
            In
          </span>
        </div>
        <div
          aria-label="Out"
          aria-sort="none"
          className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="columnheader"
          style={
            Object {
              "WebkitFlex": "0 1 130px",
              "flex": "0 1 130px",
              "msFlex": "0 1 130px",
            }
          }
          tabIndex={0}
        >
          <span
            className="ReactVirtualized__Table__headerTruncatedText"
            title="Out"
          >
            Out
          </span>
        </div>
      </div>
      <div
        aria-label="grid"
        aria-readonly={true}
        className="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
        onScroll={[Function]}
        role="rowgroup"
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "boxSizing": "border-box",
            "direction": "ltr",
            "height": -25,
            "overflowX": "hidden",
            "overflowY": "auto",
            "position": "relative",
            "width": 0,
            "willChange": "transform",
          }
        }
        tabIndex={0}
      />
    </div>
  </div>
</div>
`;
